### ISSUE TITLE
build fails on type error, expecting int in ...src/rdb_protocol/datum.cc

### ISSUE THREAD

**Issue by gigatexal** [rethinkdb/rethinkdb#4264]
trying to make on fedora22 rethinkdb from the git source on the build from source page

```
make: Entering directory '/home/gigatexal/rethinkdb/rethinkdb'
    [1/203] CC build/release/obj/rdb_protocol/datum.o
src/rdb_protocol/datum.cc: In function ‘void ql::debug_print(printf_buffer_t*, const ql::datum_t&)’:
src/rdb_protocol/datum.cc:2230:83: error: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘ql::datum_t::internal_type_t’ [-Werror=format=]
         buf->appendf("datum/garbage{internal_type=%d}", d.data.get_internal_type());
                                                                                   ^
cc1plus: all warnings being treated as errors
src/build.mk:401: recipe for target 'build/release/obj/rdb_protocol/datum.o' failed
make[1]: *** [build/release/obj/rdb_protocol/datum.o] Error 1
```

environment specific stuff:

```
[gigatexal@localhost rethinkdb]$ cpp -v
Using built-in specs.
COLLECT_GCC=cpp
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,objc,obj-c++,fortran,ada,go,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --disable-libgcj --with-default-libstdcxx-abi=c++98 --with-isl --enable-libmpx --enable-gnu-indirect-function --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
gcc version 5.1.1 20150422 (Red Hat 5.1.1-1) (GCC) 
COLLECT_GCC_OPTIONS='-E' '-v' '-mtune=generic' '-march=x86-64'
 /usr/libexec/gcc/x86_64-redhat-linux/5.1.1/cc1 -E -quiet -v - -mtune=generic -march=x86-64
ignoring nonexistent directory "/usr/lib/gcc/x86_64-redhat-linux/5.1.1/include-fixed"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-redhat-linux/5.1.1/../../../../x86_64-redhat-linux/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/gcc/x86_64-redhat-linux/5.1.1/include
 /usr/local/include
 /usr/include
End of search list.
```

**Comment by mlucy** [comment]
@AtnNn, any insight?

**Comment by AtnNn** [comment]
It looks like GCC 5.1 warns about formatting an `enum class` as `%d`.

**Comment by gigatexal** [comment]
Is it a warning or a hard failure. Can I pass a flag to ignore this?

Sent from Outlook




On Fri, May 22, 2015 at 7:09 PM -0700, "Etienne Laurin" <notifications@github.com> wrote:












It looks like GCC 5.1 warns about formatting an enum class as %d.

**Comment by AtnNn** [comment]
@gigatexal Are you able to compile with `make ALLOW_WARNINGS=1`?

**Comment by gigatexal** [comment]
I'll try that when I get back to my desk 

Sent from Outlook




On Fri, May 22, 2015 at 7:10 PM -0700, "Etienne Laurin" <notifications@github.com> wrote:












@gigatexal Are you able to compile with make ALLOW_WARNINGS=1?

**Comment by gigatexal** [comment]
that seemed to work. 

@localhost rethinkdb]$ sudo make ALLOW_WARNINGS=1
[sudo] password for gigatexal: 
    [1/203] CC build/release/obj/rdb_protocol/datum.o
src/rdb_protocol/datum.cc: In function ‘void ql::debug_print(printf_buffer_t*, const ql::datum_t&)’:
src/rdb_protocol/datum.cc:2230:83: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘ql::datum_t::internal_type_t’ [-Wformat=]
         buf->appendf("datum/garbage{internal_type=%d}", d.data.get_internal_type());
                                                                                   ^
    [2/203] CC build/release/obj/rdb_protocol/protocol.o
src/rdb_protocol/protocol.cc: In instantiation of ‘void rdb_r_unshard_visitor_t::unshard_range_batch(const query_t&, sorting_t) [with query_response_t = rget_read_response_t; query_t = intersecting_geo_read_t]’:
src/rdb_protocol/protocol.cc:656:74:   required from here
src/rdb_protocol/protocol.cc:761:30: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘ql::skey_version_t’ [-Wformat=]
                     strprintf("INTERNAL ERROR: mismatched skey versions %d and %d.",
                              ^
src/rdb_protocol/protocol.cc:761:30: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘ql::skey_version_t’ [-Wformat=]
src/rdb_protocol/protocol.cc: In instantiation of ‘void rdb_r_unshard_visitor_t::unshard_range_batch(const query_t&, sorting_t) [with query_response_t = rget_read_response_t; query_t = rget_read_t]’:
src/rdb_protocol/protocol.cc:718:61:   required from here
src/rdb_protocol/protocol.cc:761:30: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘ql::skey_version_t’ [-Wformat=]
src/rdb_protocol/protocol.cc:761:30: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘ql::skey_version_t’ [-Wformat=]

**Comment by anatol** [comment]
We see the same issue on Arch. It is related to GCC 5.1. http://pkgbuild.com/~heftig/logs3/rethinkdb/rethinkdb-2.0.2-1-x86_64-build.log

What is recommended solution? Do I need to use `ALLOW_WARNINGS=1`?

**Comment by AtnNn** [comment]
@anatol: yes, that is the recommended solution

**Comment by anatol** [comment]
Thanks.

I added it but I still see compile issues on Arch:

```
    [119/376] CC build/release/obj/serializer/log/data_block_manager.o
    [120/376] CC build/release/obj/serializer/log/metablock_manager.o
In file included from ./src/arch/types.hpp:10:0,
                 from ./src/serializer/log/data_block_manager.hpp:7,
                 from src/serializer/log/data_block_manager.cc:2:
src/serializer/log/data_block_manager.cc: In member function 'size_t data_block_manager_t::compute_gc_concurrency() const':
./src/errors.hpp:25:70: error: the value of 'GC_HIGH_RATIO' is not usable in a constant expression
 #define CT_ASSERT(e) do { enum { compile_time_assert_error = 1/(!!(e)) }; } while (0)
                                                                      ^
src/serializer/log/data_block_manager.cc:883:5: note: in expansion of macro 'CT_ASSERT'
     CT_ASSERT(GC_HIGH_RATIO > GC_START_RATIO);
     ^
src/serializer/log/data_block_manager.cc:47:14: note: 'GC_HIGH_RATIO' was not declared 'constexpr'
 const double GC_HIGH_RATIO = 0.5;
              ^
In file included from ./src/arch/types.hpp:10:0,
                 from ./src/serializer/log/data_block_manager.hpp:7,
                 from src/serializer/log/data_block_manager.cc:2:
./src/errors.hpp:25:70: error: the value of 'GC_HIGH_RATIO' is not usable in a constant expression
 #define CT_ASSERT(e) do { enum { compile_time_assert_error = 1/(!!(e)) }; } while (0)
                                                                      ^
src/serializer/log/data_block_manager.cc:883:5: note: in expansion of macro 'CT_ASSERT'
     CT_ASSERT(GC_HIGH_RATIO > GC_START_RATIO);
     ^
src/serializer/log/data_block_manager.cc:47:14: note: 'GC_HIGH_RATIO' was not declared 'constexpr'
 const double GC_HIGH_RATIO = 0.5;
              ^
In file included from ./src/arch/types.hpp:10:0,
                 from ./src/serializer/log/data_block_manager.hpp:7,
                 from src/serializer/log/data_block_manager.cc:2:
./src/errors.hpp:25:70: error: enumerator value for 'compile_time_assert_error' is not an integer constant
 #define CT_ASSERT(e) do { enum { compile_time_assert_error = 1/(!!(e)) }; } while (0)
                                                                      ^
src/serializer/log/data_block_manager.cc:883:5: note: in expansion of macro 'CT_ASSERT'
     CT_ASSERT(GC_HIGH_RATIO > GC_START_RATIO);
     ^
./src/errors.hpp:25:70: error: the value of 'GC_START_RATIO' is not usable in a constant expression
 #define CT_ASSERT(e) do { enum { compile_time_assert_error = 1/(!!(e)) }; } while (0)
                                                                      ^
src/serializer/log/data_block_manager.cc:884:5: note: in expansion of macro 'CT_ASSERT'
     CT_ASSERT(GC_START_RATIO > GC_STOP_RATIO);
     ^
src/serializer/log/data_block_manager.cc:42:14: note: 'GC_START_RATIO' was not declared 'constexpr'
 const double GC_START_RATIO = 0.15;
              ^
In file included from ./src/arch/types.hpp:10:0,
                 from ./src/serializer/log/data_block_manager.hpp:7,
                 from src/serializer/log/data_block_manager.cc:2:
./src/errors.hpp:25:70: error: the value of 'GC_START_RATIO' is not usable in a constant expression
 #define CT_ASSERT(e) do { enum { compile_time_assert_error = 1/(!!(e)) }; } while (0)
                                                                      ^
src/serializer/log/data_block_manager.cc:884:5: note: in expansion of macro 'CT_ASSERT'
     CT_ASSERT(GC_START_RATIO > GC_STOP_RATIO);
     ^
src/serializer/log/data_block_manager.cc:42:14: note: 'GC_START_RATIO' was not declared 'constexpr'
 const double GC_START_RATIO = 0.15;
              ^
In file included from ./src/arch/types.hpp:10:0,
                 from ./src/serializer/log/data_block_manager.hpp:7,
                 from src/serializer/log/data_block_manager.cc:2:
./src/errors.hpp:25:70: error: enumerator value for 'compile_time_assert_error' is not an integer constant
 #define CT_ASSERT(e) do { enum { compile_time_assert_error = 1/(!!(e)) }; } while (0)
                                                                      ^
src/serializer/log/data_block_manager.cc:884:5: note: in expansion of macro 'CT_ASSERT'
     CT_ASSERT(GC_START_RATIO > GC_STOP_RATIO);
     ^
src/build.mk:396: recipe for target 'build/release/obj/serializer/log/data_block_manager.o' failed
make[1]: *** [build/release/obj/serializer/log/data_block_manager.o] Error 1
make[1]: *** Waiting for unfinished jobs....
```

**Comment by anatol** [comment]
I wonder why gcc 5.1 thinks that `GC_HIGH_RATIO` is not a constant. Is it a compiler bug?

**Comment by AtnNn** [comment]
@anatol variables used in compile-time expressions should be constexpr, not const.

Here's a preliminary patch: 242a351

**Comment by anatol** [comment]
@AtnNn Thank you for this quick reply. It helped me to build rethinkdb with gcc 5.1. Everything looks fine so far.

**Comment by danielmewes** [comment]
@AtnNn can you put that patch up for CR or does this require additional work?

**Comment by AtnNn** [comment]
Fixed in d459f6d
